### DISCUSSION TITLE
How to set page labels?

### DISCUSSION THREAD

**Discussion by cmahnke** [py-pdf/fpdf2#1138] [General]
It's possible to add custom labels to pages, that are for example page numbers in different schemes, like Roman numerals.
Is this possible with `fpdf2`? There is a python module which should be able to do this:
https://github.com/lovasoa/pagelabels-py

**Comment by Lucas-C** [comment]
Hi @cmahnke

You can easily add page numbers at the bottom of pages using the `.footer()` & `.page_no()` methods, as described in our tutorial:
https://py-pdf.github.io/fpdf2/Tutorial.html#tuto-2-header-footer-page-break-and-image

And if you want to convert numbers to roman numerals, you can use the [`roman` pip package](https://pypi.org/project/roman/), or code your own conversion function: https://www.geeksforgeeks.org/python-program-to-convert-integer-to-roman/

**Reply by cmahnke** [reply]
Thanks for your reply, but that's not what I was trying to ask:
I was asking for Page labels in terms of Section 12.4.2 of the [specification](https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/PDF32000_2008.pdf)

If anyone has the same problem in the future: one can solve this using [`pikepdf`](https://pikepdf.readthedocs.io/en/latest/api/models.html#pikepdf.NumberTree)

**Comment by gmischler** [comment]
I've checked out the relevant parts of the specs, and this looks like it would be a very helpful feature.
It essentially decouples logical page numbers from physical page numbers, allowing to have several different sections in a document following their own page numbering schemes.
For example, a section with a multi-page TOC could have its own page numbers (say using i, ii, iii, etc.), while the content part of the document keeps the conventional arabic numbers (1, 2, 3, etc.). Our current string substitution hack for inserting page numbers would then become obsolete. Any conforming reader would also recognize those custom page labels.
Any takers?
